lms/models/application_settings.py

```diff
@@ -100,6 +102,42 @@ def get_secret(self, aes_service, group, key) -> Optional[str]:
 
         return aes_service.decrypt(aes_iv, aes_value).decode("utf-8")
 
+    @classmethod
+    def matching(cls, column: InstrumentedAttribute, spec: dict):
+        """
+        Return a clause to filter an SQLAlchemy query.
+
+        This method accepts a match dict which should have keys like this:
+
+            {
+                # Specify an exact key should have an exact value
+                "group.key": "exact_value_to_match"
+                # Specify an exact key should exist
+                "group.key": ...
+                # Specify a group should exist
+                "group": ...
+            }
+
+        :param column: The column to apply this to
+        :param spec: The filter specification as described above
+        """
+        clauses = []
+
+        for joined_key, value in spec.items():
+            group, key = (
+                joined_key.split(".") if "." in joined_key else (joined_key, None)
+            )
+
+            if key is None:
+                clauses.append(column.has_key(group))
+
+            elif value is not ...:
+                clauses.append(column.contains({group: {key: value}}))
+            else:
+                clauses.append(column[group].has_key(key))
```

Comment:
This total randomness of the interface is unpleasant. In SQLAlchemy 2.0 there is a `path_exists` function which would unify the "group" and "group.key" situations.

I wanted to hide this away, but also to have it be re-usable. Not sure if it's worth it or not.